Code/Yospace/Source/YospaceNotificationsHandler.swift

```diff
@@ -0,0 +1,212 @@
+        guard let currentAd: YOAdvert = self.currentAd,
+              let ad: THEOplayerSDK.Ad = self.adsMap[currentAd],
+              progressEventsList.contains(name) else { return }
+        let remaining: Double = currentAd.remainingTime(self.player.currentTime)
+        let duration: Double = currentAd.duration
+        let progress: Double = (duration - remaining) / duration
+        // avoid reporting updates when skipping; values should be in the range of [0.25, 0.75]
+        if progress < 1 {
+            self.adIntegrationController.updateAdProgress(ad: ad, progress: progress)
+        }
+    }
+
+    @objc private func analyticUpdateDidOccur(notification: NSNotification) {
+        for adBreak: YOAdBreak in self.session.adBreaks(.linearType) as! Array<YOAdBreak>  {
+            self.processAdBreak(yospaceAdBreak: adBreak)
+        }
+        for adBreak: YOAdBreak in self.session.adBreaks(.nonLinearType) as! Array<YOAdBreak> {
+            self.processAdBreak(yospaceAdBreak: adBreak)
+        }
+        for adBreak: YOAdBreak in self.session.adBreaks(.displayType) as! Array<YOAdBreak> {
+            self.processAdBreak(yospaceAdBreak: adBreak)
+        }
+    }
+
+    @objc private func sessionErrorDidOccur(notification: Notification) {
+        let info = notification.userInfo
+        let code: NSNumber? = info?[YOErrorCodeKey] as? NSNumber
+
+        if code == NSNumber(value: YOSessionError.sessionTimeout.rawValue) {
+            let error = YospaceError.error(msg: YOSessionError.sessionTimeout.errorMessage)
+            self.adIntegrationController.error(error: error)
+            self.session.shutdown()
+        } else if code == NSNumber(value: YOSessionError.unresolvedBreak.rawValue) {
+            let error = YospaceError.error(msg: YOSessionError.unresolvedBreak.errorMessage)
+            self.adIntegrationController.error(error: error)
+        } else if code == NSNumber(value: YOSessionError.parseError.rawValue),
+                  let errors: Array<YOParsingError> = self.session.parsingErrors() as? Array<YOParsingError> {
+            errors.forEach { parsingError in
+                let error = YospaceError.error(msg: YOSessionError.parseError.errorMessage + "Parsing error: \(parsingError.toJsonString())")
+                self.adIntegrationController.error(error: error)
+            }
+        }
+    }
+
+    @objc private func adBreakDidEndEarly(notification: Notification) {
+        if let currentAd: YOAdvert = self.currentAd,
+           let ad: THEOplayerSDK.Ad = self.adsMap[currentAd] {
+            self.adIntegrationController.skipAd(ad: ad)
+            self.currentAd = nil
+        }
+        self.adBreakDidEnd(notification: notification)
+    }
+
+    @objc private func trackingErrorDidOccur(notification: Notification) {
+        let info = notification.userInfo
+        let trackingError: YOTrackingError = info?[YOTrackingErrorKey] as! YOTrackingError
```

Comment:
if the `info` is optional, better not to force unwrap the values.

Reply:
fixed